File: samplecode/perl/naive.pl
#!/usr/bin/env perl

#   Copyright 2014 athenahealth, Inc.
#
#   Licensed under the Apache License, Version 2.0 (the "License"); you
#   may not use this file except in compliance with the License.  You
#   may obtain a copy of the License at
#
#       http://www.apache.org/licenses/LICENSE-2.0
#
#   Unless required by applicable law or agreed to in writing, software
#   distributed under the License is distributed on an "AS IS" BASIS,
#   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
#   implied.  See the License for the specific language governing
#   permissions and limitations under the License.

use strict;
use warnings;

use LWP;
use JSON;
use Data::Dumper;
use URI::Escape;

# Setup all the constants
my $practiceid = 000000;
my $version = 'preview1';
my $key = 'CHANGEME: YOUR_API_KEY';
my $secret = 'CHANGEME: YOUR_API_SECRET';
my $authurl = 'https://api.athenahealth.com/oauthpreview/token';

# Make a new user agent to make requests through
my $useragent = LWP::UserAgent->new();

# Authentication parameters and headers
my $parameters = 'grant_type=client_credentials';
my $headers = HTTP::Headers->new();
$headers->authorization_basic($key, $secret);

# Make a new POST request, set the parameters, headers, and content-type
my $req = HTTP::Request->new('POST', $authurl, $headers, $parameters);
$req->content_type('application/x-www-form-urlencoded');

# Prepare and send
$useragent->prepare_request($req);
my $res = $useragent->send_request($req);

# Decode from JSON and save the token for later
my $decoded = decode_json($res->content);
my $token = $decoded->{'access_token'};

print "$token\n";

my $baseurl = 'https://api.athenahealth.com/';


# GET /customfields
my $url2 = $baseurl . $version . '/' . $practiceid . '/customfields';

# Add in the Authorization header with the token
my $req2 = HTTP::Request->new('GET', $url2);
$req2->header('Authorization' => 'Bearer ' . $token);
$useragent->prepare_request($req2);
print(Dumper $req2);

my $res2 = $useragent->send_request($req2);
my $customfields = decode_json($res2->content);
print Dumper($customfields);


# GET /departments
my %params = (
	'limit' => 1,
);
	
# GETs take parameters in the URL.  This escapes and joins everything properly.
my $query = join('&', map { 
	uri_escape($_) . '=' . uri_escape($params{$_}) 
} keys(%params));

if ($query) {
	$query = '?' . $query;
}

my $url3 = $baseurl . $version . '/' . $practiceid . '/departments';
my $req3 = HTTP::Request->new('GET', $url3 . $query);
$req3->header('Authorization' => 'Bearer ' . $token);
$useragent->prepare_request($req3);
my $res3 = $useragent->send_request($req3);

my $departments = decode_json($res3->content);

File: samplecode/ruby/naive.rb
#!/usr/bin/env ruby

#    Copyright 2014 athenahealth, Inc.
#
#   Licensed under the Apache License, Version 2.0 (the "License"); you
#   may not use this file except in compliance with the License.  You
#   may obtain a copy of the License at
#
#       http://www.apache.org/licenses/LICENSE-2.0
#
#   Unless required by applicable law or agreed to in writing, software
#   distributed under the License is distributed on an "AS IS" BASIS,
#   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
#   implied.  See the License for the specific language governing
#   permissions and limitations under the License.

require 'net/https'
require 'uri'
require 'cgi'
require 'rubygems'
require 'json'
require 'pp'

key = 'CHANGEME: YOUR_API_KEY'
secret = 'CHANGEME: YOUR_API_SECRET'
version = 'preview1'
practiceid = 000000


# Start a connection
uri = URI.parse('https://api.athenahealth.com/')
connection = Net::HTTP.new(uri.host, uri.port)
connection.use_ssl = true

# Monkey patch to make Net::HTTP do proper SSL verification.
# Background reading: 
# http://stackoverflow.com/a/9238221
# http://blog.spiderlabs.com/2013/06/a-friday-afternoon-troubleshooting-ruby-openssl-its-a-trap.html
def connection.proper_ssl_context!
  ssl_context = OpenSSL::SSL::SSLContext.new
  ssl_context.verify_mode = OpenSSL::SSL::VERIFY_PEER
  cert_store = OpenSSL::X509::Store.new
  cert_store.set_default_paths
  ssl_context.cert_store = cert_store
  @ssl_context = ssl_context
end
connection.proper_ssl_context!
# End monkey patch

# Authenticate (basic access authentication)
auth_paths = {
  'v1' => 'oauth',
  'preview1' => 'oauthpreview',
  'openpreview1' => 'oauthopenpreview',
}
request = Net::HTTP::Post.new("/#{auth_paths[version]}/token")
request.basic_auth(key, secret)
request.set_form_data({'grant_type' => 'client_credentials'})

response = connection.request(request)
authorization = JSON.parse(response.body)

# Save the token for later
token = authorization['access_token']
puts token


# This makes making paths easier
base_path = "/#{version}/#{practiceid}"


# GET /departments
params = {
  'limit' => 1,
}
query = params.map { |k,v| [k,v].map { |x| CGI.escape(x.to_s) }.join('=') }.join('&')
if query
  query = '?' + query
end

req2 = Net::HTTP::Get.new("#{base_path}/departments#{query}")
req2['authorization'] = "Bearer #{token}"

res2 = connection.request(req2)
departments = JSON.parse(res2.body)

puts 'Departments:'
pp departments


# POST /appointments/{appointmentid}/notes
appt_id = 1
req3 = Net::HTTP::Post.new("#{base_path}/appointments/#{appt_id}/notes")
req3.set_form_data({'notetext' => 'Notes from Ruby'})
req3['authorization'] = "Bearer #{token}"

res3 = connection.request(req3)
note = JSON.parse(res3.body)

puts 'Note:'
pp note
